### PR TITLE
Building libcxx with KLEE.

### PR DESCRIPTION
**Pull request by rsprabery**
More work is needed to get things like re2 to run. The `klee_driver.cpp`
needs to be aware of size limits.

@laurentsimon PTAL.

### CODE REVIEW

File: fuzzers/klee/builder.Dockerfile [hunk 1]
```diff
@@ -49,38 +49,53 @@ RUN git clone https://github.com/stp/stp.git /stp && \
...
-# Install KLEE. Use my personal repo containing additional scripts we need for now.
+
+# Install KLEE. Use my personal repo containing additional scripts we need for
+# now.
 RUN git clone https://github.com/lmrs2/klee.git /klee && \
     cd /klee && \
-    git checkout debug && \
+    git checkout debug
+
+# the libcxx build script in the KLEE repo depends on wllvm:
```
Comment (jonathanmetzman) [inline fuzzers/klee/builder.Dockerfile:65]: nit: capitalize first letter in comments please.

File: fuzzers/klee/fuzzer.py [hunk 2]
```diff
@@ -38,7 +38,8 @@ def prepare_build_environment():
     os.environ['CC'] = 'gclang'
     os.environ['CXX'] = 'gclang++'
 
-    os.environ['FUZZER_LIB'] = '/libAFL.a -L/ -lKleeMock'
+    os.environ['FUZZER_LIB'] = '/libAFL.a -L/ -lKleeMock -lpthread'
```
Comment (jonathanmetzman) [inline fuzzers/klee/fuzzer.py:41]: I think `-pthread` is set by default in CXXFLAGS and CFLAGS.
Is it actually needed?

File: fuzzers/klee/fuzzer.py [hunk 3]
```diff
@@ -66,13 +67,23 @@ def emptydir(path):
...
+    """Run the command |command|, optionally, run |ulimit_cmd| first."""
     cmd = ' '.join(command)
     print('[run_cmd] {}'.format(cmd))
 
     output_stream = subprocess.DEVNULL if hide_output else None
-    ret = subprocess.call(command, stdout=output_stream, stderr=output_stream)
+    if ulimit_cmd:
+        ulimit_command = [ulimit_cmd + ';']
+        ulimit_command.extend(command)
+        ret = subprocess.call(' '.join(ulimit_command),
```
Comment (jonathanmetzman) [inline fuzzers/klee/fuzzer.py:79]: Don't think the join call is needed. `call` should work on a list.

File: fuzzers/klee/fuzzer.py [hunk 4]
```diff
@@ -66,13 +67,23 @@ def emptydir(path):
     os.mkdir(path)
 
 
-def run(command, hide_output=False):
-    """Run a the command |command|"""
+def run(command, hide_output=False, ulimit_cmd=None):
```
Comment (jonathanmetzman) [inline fuzzers/klee/fuzzer.py:70]: Why is this the ulimit stuff needed?

Comment (kren1) [reply fuzzers/klee/fuzzer.py:70]: STP uses recursion when constraint solving. If you have a stack limit KLEE reaches it quite quickly and dies. However I'm not sure if these work inside docker, usually you have to tell docker, that the container has no stack limit.

File: fuzzers/klee/fuzzer.py (hunk 2 above)
Comment (rsprabery) [reply fuzzers/klee/fuzzer.py:41]: It was needed to resolve an issue with libpng, though the rest of that fix is missing in this PR.

File: fuzzers/klee/fuzzer.py (hunk 4 above)
Comment (rsprabery) [reply fuzzers/klee/fuzzer.py:70]: As @kren1 mentioned, this is necessary for STP. 

I've tested this with both libpng and re2 (fixes for those specific benchmarks coming in another PR) and this fixed the issue.

I had set the unlimit in the dockerfile, but it didn't appear to take. This was the only way I was able to clear up the issue with STP. Open to suggestions if anyone has a cleaner way to do this.

**Comment by laurentsimon** [comment]
FYI, AFL also needs this, but does it itself in afl-fuzz.c
<https://github.com/google/AFL/blob/master/afl-fuzz.c#L2023> I don't think
KLEE has a similar option, does it?

>

File: fuzzers/klee/fuzzer.py (hunk 3 above)
Comment (rsprabery) [reply fuzzers/klee/fuzzer.py:79]: I double checked, in general, call works with lists as you said - but I'm getting an issue where the side effects of ulimit -s aren't visible to klee (increased stack size) unless I do this join.

**Comment by rsprabery** [comment]
Thanks for the feedback! I've addressed the comments. I think the ulimit thing I've got going here is a temporary fix for now. We'll want to do something to make that cleaner. Maybe add an option to KLEE itself?

**Comment by jonathanmetzman** [comment]
Didn't bother checking that presubmit wasn't one of the failures. Could you fix the please? (no rush)

**Pull request merged**